### ISSUE TITLE
(dgst+hmac) MAC parameter error "key:123456901234567890123456789012"

### ISSUE THREAD

**Issue by vt-alt** [openssl/openssl#10814]
While testing [gost-engine](https://github.com/gost-engine/engine/pull/187#issuecomment-573364326) I've stumbled upon this issue with openssl master branch:

```
openssl (master)$ apps/openssl dgst -sha256 -mac hmac -macopt key:123456901234567890123456789012 </dev/null
MAC parameter error "key:123456901234567890123456789012"

openssl (master)$ git status
On branch master
Your branch is up-to-date with 'origin/master'.
nothing to commit, working tree clean

openssl (master)$ git describe
OpenSSL_1_1_1-2428-ge0e68f9e34
openssl (master)$
```

**Comment by levitte** [comment]
Yup, I can see why this happens, an think I know how to deal with it.  I'll be back.

**Comment by slontis** [comment]
Multiple things are wrong..
(1) It looks like if mac_name is "hmac" then it needs to tack the digest md onto the macopts
i.e apps/openssl dgst -sha256 -mac hmac -macopt md:sha256 -macopt key:1234569012345678901234567890123 will get you a little further..

(2) It then tries to do EVP_PKEY_keygen(mac_ctx, ).. which falls over.. (Not sure what that is trying to do).

There is a openssl mac application that will work for HMAC that only uses -macopt..

**Comment by levitte** [comment]
No, @slontis. The issue is that the ctrl_str that's being called is not coded correctly, as it will not pass command names that libcrypto doesn't recognise.

**Comment by levitte** [comment]
I found out why this happens...  It's the case of raw keys that's not properly intercepted in `pkey_mac_str_ctrl()`.  We really shouldn't need to intercept that, but since it has local implications, we must.

**Comment by levitte** [comment]
It turns out that my analysis was incorrect...  or well, partly incorrect.  #10947 solves the general problem of passing arbitrary string controls to provider implementations, but the error persists, so I did a bit of debugging, and it turns out that the interaction between the EVP_PKEY_METHOD \<-\> EVP_MAC bridge `crypto/evp/pkey_mac.c` handles the HMAC case differently than in 1.1.1.

In 1.1.1, the HMAC EVP_PKEY_METHOD makes a local copy of the key and uses it later, while the bridge passes it on to the HMAC EVP_MAC, which passes it immediately to the low-level functions, and because the MD hasn't been set yet for that MAC at this point (that happens during the `EVP_DigestSignInit` call), the HMAC functions go belly up.

**Comment by vasavenk** [comment]
hey @levitte ! trying to navigate through this. I am still getting an error when doing openssl sha256 -mac HMAC -macopt hexkey:$keybytes .... were you able to figure it out/have any recommendations/workarounds?

**Comment by nhorman** [comment]
@levitte is this still an issue being investigated?

**Comment by levitte** [comment]
Er.....  I think we need to investigate again to see if we fixed this.

**Comment by nhorman** [comment]
ack, moving to consideration
